### PR TITLE
I have been missing a timeout to devices over TCP connection. This is an attempt at implementing this functionality.

### PR DESCRIPTION
**Pull request by farthornas**
The code for tcp connection has been tested extensively to adb devices and have worked without any known issues, however the unittests for tcp connection could need input which would be much appreciated.

### CODE REVIEW

File: adb/adb_commands.py [hunk 1]
```diff
@@ -68,7 +68,7 @@ def ConnectDevice(
     used instead of a USB connection.
     """
     if serial and b':' in serial:
-        handle = common.TcpHandle(serial)
+        handle = common.TcpHandle(serial,timeout_ms=default_timeout_ms)
```
Comment (fahhem) [inline adb/adb_commands.py]: nit: space after comma

File: adb/common.py [hunk 2]
```diff
@@ -276,12 +277,13 @@ def FindDevices(cls, setting_matcher, device_matcher=None,
         yield handle
 
 class TcpHandle(object):
-  """TCP connection object.
+  """Tcp connection object.
```
Comment (fahhem) [inline adb/common.py]: TCP in the comments, we title-case words in classes for other reasons.

File: adb/common.py [hunk 3]
```diff
@@ -292,22 +294,34 @@ def __init__(self, serial):
     else:
       host = serial
       port = 5555
+    self._timeout_ms = timeout_ms
```
Comment (fahhem) [inline adb/common.py]: set twice?

File: adb/common.py [hunk 4]
```diff
@@ -292,22 +294,34 @@ def __init__(self, serial):
...
     self._serial_number = '%s:%s' % (host, port)
-
-    self._connection = socket.create_connection((host, port))
+    self._timeout_ms = float(timeout_ms)
+    timeout = float(timeout_ms)/1000.0 if timeout_ms else None
```
Comment (fahhem) [inline adb/common.py]: self._timeout_ms / 1000.0 if timeout_ms else None

File: adb/common.py [hunk 5]
```diff
@@ -292,22 +294,34 @@ def __init__(self, serial):
     else:
       host = serial
       port = 5555
+    self._timeout_ms = timeout_ms
     self._serial_number = '%s:%s' % (host, port)
-
-    self._connection = socket.create_connection((host, port))
+    self._timeout_ms = float(timeout_ms)
+    timeout = float(timeout_ms)/1000.0 if timeout_ms else None 
+    self._connection = socket.create_connection((host,port),timeout=timeout)
```
Comment (fahhem) [inline adb/common.py]: space after comma

File: adb/common.py [hunk 6]
```diff
@@ -292,22 +294,34 @@ def __init__(self, serial):
...
+    if timeout:
+        self._connection.setblocking(0)
 
   @property
   def serial_number(self):
     return self._serial_number
 
   def BulkWrite(self, data, timeout=None):
+      if self._timeout_ms:
+          return self._connection.send(data)
```
Comment (fahhem) [inline adb/common.py]: what's the difference here? What if we did `send` in the other case too?

File: test/common_stub.py [hunk 7]
```diff
@@ -53,3 +56,43 @@ def ExpectRead(self, data):
 
   def Timeout(self, timeout_ms):
     return timeout_ms if timeout_ms is not None else self.timeout_ms
+
+class StubTcp(StubUsb):
+
+  def __signal_handler(self, signum, frame):
```
Comment (fahhem) [inline test/common_stub.py]: prefer single underscores when double isn't necessary.

File: adb/common.py (hunk 6 above)
Comment (farthornas) [reply adb/common.py]: My initial concern was in that self._connection.sendall(data) requires a blocking connection as sendall is supposed to send all data . Connection is set to non blocking (self._connection.setblocking(0)) to allow for timeout on the connection. From these points it seems self._connection.send() is the way to go, however the docs on "send()" says: "Applications are responsible for checking that all data has been sent": From what I gather (excuse my ignorance) it seems such a check is not done (adb_protocol.Send(usb, timeout)).

File: adb/common.py [hunk 8]
```diff
@@ -293,21 +295,36 @@ def __init__(self, serial):
...
+    if timeout:
+        self._connection.setblocking(0)
 
   @property
   def serial_number(self):
     return self._serial_number
 
   def BulkWrite(self, data, timeout=None):
-      return self._connection.sendall(data)
+      t = self.Timeout(timeout)/1000.0 if self.Timeout(timeout) else None
```
Comment (fahhem) [inline adb/common.py]: please fix the indentation here to 2 spaces

File: adb/common.py [hunk 9]
```diff
@@ -293,21 +295,36 @@ def __init__(self, serial):
...
+      writeable = select.select([], [self._connection], [], t)
```
Comment (fahhem) [inline adb/common.py]:
```py
_, writeable, _ = select.select([], [self._connection], [], t)
if writeable:
  return self._connection.send(data)
```

File: adb/common.py [hunk 10]
```diff
@@ -293,21 +295,36 @@ def __init__(self, serial):
...
+      if writeable[1]:
+        return self._connection.send(data)
+      msg = 'Sending data to {} timed out after {}s. No was data sent.'.format(
+              self.serial_number, t)
+      raise usb_exceptions.TcpTimeoutException(msg) 
 
   def BulkRead(self, numbytes, timeout=None):
-      return self._connection.recv(numbytes)
+      t = self.Timeout(timeout)/1000.0 if self.Timeout(timeout) else None
+      readable = select.select([self._connection], [], [], t)
```
Comment (fahhem) [inline adb/common.py]:
```py
readable, _, _ = ...
if readable:
  ...
```

File: adb/common.py [hunk 11]
```diff
@@ -293,21 +295,36 @@ def __init__(self, serial):
...
+      writeable = select.select([], [self._connection], [], t)
+      if writeable[1]:
+        return self._connection.send(data)
+      msg = 'Sending data to {} timed out after {}s. No was data sent.'.format(
+              self.serial_number, t)
+      raise usb_exceptions.TcpTimeoutException(msg) 
 
   def BulkRead(self, numbytes, timeout=None):
-      return self._connection.recv(numbytes)
+      t = self.Timeout(timeout)/1000.0 if self.Timeout(timeout) else None
```
Comment (fahhem) [inline adb/common.py]: Create a `TimeoutSeconds` function to encapsulate both of these lines:
```py
def TimeoutSeconds(self, timeout_ms):
  timeout = self.Timeout(timeout_ms)
  return timeout / 1000.0 if timeout is not None else timeout
```

**Pull request merged**